rfcs/20190610-standartizing-composite_ops.md

```diff
@@ -0,0 +1,325 @@
+signal to the processing/executing library that this function implements a
+“standard” function. The consumer library then can choose to process it
+differently should they choose to. The important part is that we define a
+standard set of attributes that users and tooling can rely on.
+
+As a start, we propose to add “implements” and “reference” as two attributes
+defining a “standard” function.
+
+For example:
+
+```
+@function.library_function(
+    implements=“google.matmul_low_rank_matrix”,
+    reference=“arxiv.org/…”)
+def matmul_lowrank(x, y):
+   “Multiplies two low rank matrices. ”
+   # Basic implementation goes here.
+```
+
+Note we use a wrapper named library_function, that under-the-hood will create a
+standard tensorflow function with standard attributes set.
+
+Arguments:
+
+* `implements`:  we propose using namespaces with the top-level namespace
+being company name, with a few reserved namespaces indicating built-in ops, 
+if so desired.  This attributes indicates what function this subgraph implements,
+and provides a hint for downstream parsers that they can replace this with 
+their own implementation should they so choose. 
+
+* `reference`:  The goal for this attribute is to provide a stable reference for the function
+implementation details. This could be a reference to non-tensorflow library (e.g. numpy.linalg.special_matmul)
+or arxiv reference. The downstream tooling can use this value as an opaque cache to verify that 
+it implements the right flavor of the function. 
+
+
+This brings us the following advantages:
+
+1) *Backward and fork compatibility*: the tooling by default will,
+ignore `implements` and just rely on function implementation available in the
+graph, thus ensuring that it “just works”. Further, if users have their
+proprietary implementations they can still ship models that would transparently
+work in open source tensorflow.
+
+2) *Easy detection*: Tools can easily detect the functions attribute and substitute with
+its own more efficient implementation if available.
+
+3) *Simplified implementations on custom hardware*: The downstream tooling can
+use provided reference implementation as a source of ground truth (that can run
+on custom hardware out of the box!) when developing custom implementation on
+their hardware.
+
+4) *Reduced implementation dependencies*: Operation maintainer can change the
+implementation without the fear of breaking the downstream tools pattern
+recognition (since there isn’t any.)
+
+5) *Forward compatibility*: Operation maintainer can even add an atomic
+implementation without breaking existing tools. No overhead in keeping the basic
+implementation available.
+
+6) *Simpler automatic conversion* As tensorflow moves forward it is conceivable
+that its computation description language (today: graph def, tomorrow:MLIR) 
+will be fully  be separated from the underlying tensorflow kernel implementations.
+This kind of standard attributes allow for more  automatic conversions
```

Comment:
Just an observation - Overall, the approach makes sense to me and is definitely a progress from where we are today. Relying on user supplied names is kind of brittle (as against a strongly typed or well defined schema) but at the same time is also very flexible.